xen/arch/arm/cpufreq/scpi_cpufreq.c

```diff
@@ -0,0 +1,353 @@
+
+    /* TODO x86 checks acpi_pstate_strict here */
+    if ( 1 && !check_freq(policy->cpu, freqs.new) )
+    {
+        printk(KERN_WARNING "Fail transfer to new freq %d\n", freqs.new);
+        return -EAGAIN;
+    }
+
+    for_each_cpu( j, &online_policy_cpus )
+        cpufreq_statistic_update(j, perf->state, next_perf_state);
+
+    perf->state = next_perf_state;
+    policy->cur = freqs.new;
+
+    return result;
+}
+
+static int scpi_cpufreq_verify(struct cpufreq_policy *policy)
+{
+    struct scpi_cpufreq_data *data;
+    struct processor_performance *perf;
+
+    if ( !policy || !(data = cpufreq_driver_data[policy->cpu] ) ||
+         !processor_pminfo[policy->cpu] )
+        return -EINVAL;
+
+    perf = &processor_pminfo[policy->cpu]->perf;
+
+    cpufreq_verify_within_limits(policy, 0,
+        perf->states[perf->platform_limit].core_frequency * 1000);
+
+    return cpufreq_frequency_table_verify(policy, data->freq_table);
+}
+
+static int scpi_cpufreq_cpu_init(struct cpufreq_policy *policy)
+{
+    unsigned int i;
+    unsigned int valid_states = 0;
+    unsigned int cpu = policy->cpu;
+    unsigned int curr_state, curr_freq;
+    struct scpi_cpufreq_data *data;
+    unsigned int result = 0;
+    struct processor_performance *perf;
+    struct scpi_dvfs_info *info;
+    int domain_id;
+    struct device *cpu_dev;
+
+    data = xzalloc(struct scpi_cpufreq_data);
+    if ( !data )
+        return -ENOMEM;
+
+    cpufreq_driver_data[cpu] = data;
+
+    data->perf = &processor_pminfo[cpu]->perf;
+
+    perf = data->perf;
+    policy->shared_type = perf->shared_type;
+
+    data->freq_table = xmalloc_array(struct cpufreq_frequency_table,
+                                    (perf->state_count + 1));
+    if ( !data->freq_table )
+    {
+        result = -ENOMEM;
+        goto err_unreg;
+    }
+
+    /* detect transition latency */
+    policy->cpuinfo.transition_latency = 0;
+    for ( i = 0; i < perf->state_count; i++ )
+    {
+        if ( (perf->states[i].transition_latency * 1000) >
+            policy->cpuinfo.transition_latency )
+            policy->cpuinfo.transition_latency =
+                perf->states[i].transition_latency * 1000;
+    }
+
+    policy->governor = cpufreq_opt_governor ? : CPUFREQ_DEFAULT_GOVERNOR;
+
+    /* table init */
+    for ( i = 0; i < perf->state_count; i++ )
+    {
+        if ( i > 0 && perf->states[i].core_frequency >=
+             data->freq_table[valid_states - 1].frequency / 1000 )
+            continue;
+
+        data->freq_table[valid_states].index = i;
+        data->freq_table[valid_states].frequency =
+            perf->states[i].core_frequency * 1000;
+        valid_states++;
+    }
+    data->freq_table[valid_states].frequency = CPUFREQ_TABLE_END;
+    perf->state = 0;
+
+    result = cpufreq_frequency_table_cpuinfo(policy, data->freq_table);
+    if ( result )
+        goto err_freqfree;
+
+    /* Fill in fields needed for frequency changing. */
+    cpu_dev = get_cpu_device(cpu);
```

Comment:
cpu_dev + domain_id + info can be retrieved at the very beginning, so if these fail then we don't even try to allocate memory and have simpler roll-back?

Reply:
Make sense. I planned to do such changes.